clusterloader2/pkg/measurement/common/ooms_tracker.go

```diff
@@ -0,0 +1,200 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+	"k8s.io/perf-tests/clusterloader2/pkg/measurement"
+	measurementutil "k8s.io/perf-tests/clusterloader2/pkg/measurement/util"
+	"k8s.io/perf-tests/clusterloader2/pkg/measurement/util/informer"
+	"k8s.io/perf-tests/clusterloader2/pkg/util"
+)
+
+const (
+	clusterOomsTrackerEnabledFlagName = "clusterOomsTrackerEnabled"
+	clusterOomsTrackerName            = "ClusterOomsTracker"
+	informerTimeout                   = time.Minute
+	oomEventReason                    = "OOMKilling"
+	oomEventMsgRegexPattern           = `Kill process (\d+) \((.+)\) score \d+ or sacrifice child\nKilled process \d+ .+ total-vm:(\d+kB), anon-rss:\d+kB, file-rss:\d+kB.*`
+)
+
+func init() {
+	if err := measurement.Register(clusterOomsTrackerName, createClusterOomsTrackerMeasurement); err != nil {
+		klog.Fatalf("Cannot register %s: %v", clusterOomsTrackerName, err)
+	}
+}
+
+func createClusterOomsTrackerMeasurement() measurement.Measurement {
+	fieldSelector := fields.Set{
+		"reason": oomEventReason,
+	}.AsSelector().String()
+	return &clusterOomsTrackerMeasurement{
+		selector: &measurementutil.ObjectSelector{
+			FieldSelector: fieldSelector,
+			Namespace:     metav1.NamespaceAll,
+		},
+		msgRegex: regexp.MustCompile(oomEventMsgRegexPattern),
+	}
+}
+
+type clusterOomsTrackerMeasurement struct {
+	// lock for synchronized access to ooms field
+	lock      sync.Mutex
+	selector  *measurementutil.ObjectSelector
+	msgRegex  *regexp.Regexp
+	isRunning bool
+	stopCh    chan struct{}
+	ooms      []oomEvent
+}
+
+type oomEvent struct {
+	Node          string    `json:"node"`
+	Process       string    `json:"process"`
```

Comment:
I'm not sure whether this will be the "process" in the descriptive way we hope to have. E.g. here is an example event.
```
- apiVersion: v1
  count: 16
  eventTime: null
  firstTimestamp: "2020-06-08T15:16:32Z"
  involvedObject:
    kind: Node
    name: gke-cluster-1-default-pool-4b0893d6-mr7x
    uid: gke-cluster-1-default-pool-4b0893d6-mr7x
  kind: Event
  lastTimestamp: "2020-06-08T15:16:47Z"
  message: |-
    (combined from similar events): Memory cgroup out of memory: Kill process 918454 (exe) score 10000 or sacrifice child
    Killed process 918454 (exe) total-vm:7972kB, anon-rss:4kB, file-rss:0kB, shmem-rss:0kB
  metadata:
    creationTimestamp: "2020-06-08T15:16:32Z"
    name: gke-cluster-1-default-pool-4b0893d6-mr7x.16169a7fe54f4e6b
    namespace: default
    resourceVersion: "8174"
    selfLink: /api/v1/namespaces/default/events/gke-cluster-1-default-pool-4b0893d6-mr7x.16169a7fe54f4e6b
    uid: da4e9029-c406-4d86-92d4-d9cb2246d837
  reason: OOMKilling
  reportingComponent: ""
  reportingInstance: ""
  source:
    component: kernel-monitor
    host: gke-cluster-1-default-pool-4b0893d6-mr7x
  type: Warning
```

The Process here will be parsed as "exe" which doesn't give us much. To be clear, I didn't name the pod/container `exe` or anything like that. 
Ideally, we'd have a pod name somewhere here, but it looks like it's not exported anywhere by node-problem-detector. I'm fine with that as long as there is an instruction documented somewhere on how to connect these OOM events with terminated pods, e.g. can we get it from systemd / kubelet logs?

Reply:
When checking for OOMs directly, I've also worked with processes, not containers/pods.
It is usually descriptive enough (pod_nanny being an exception).
Documenting how to debug it further would be nice (probably by checking kubelet logs (via log viewer) for restarts).

Reply:
I put a small Markdown entry about the measurement in the newest version of this PR - PTAL.

@mm4tt @jkaniuk shall I get rid of the "Process" field altogether then? If so, I think we can leave out "ProcessMemory" field as well as it's not the actual memory limit enforced on the pod but total virtual memory used up by the process. I would keep the rest of the `oomEvent` struct fields, though - "Node" is essential for narrowing the search for essential information in logs, "Pid" may be convenient for grepping and "Time" is also convenient for having a time reference when the OOM was recorded.

Reply:
"check `kern.log` as well as `systemd.log` files of an appropriate node for the name of OOMing pod/container or the pod's memory limit, for instance."

Did you verify that this info is there? Do we need to check both files?

Process is quite important, please leave it as it is.
If ProcessMemory is the same value as reported in kern.log, it was handy for me in the past and did match container limit (exceeding it a little bit).
I would not remove any of those fields.

Reply:
Actually, `systemd.log` duplicates information from `kern.log`. Moreover, it contains an addition of logs from kubelet and container runtime regarding the OOMing container/pod, hence I'll just leave `systemd.log` as the suggested file to look at in the doc.

Reply:
I really hope that in not that far future once we migrate to new events API, the relatedObject will contain all the information about the Pod itself.
Might be useful to add a comment to evaluate if we can get more once NPD is migrated to new events (hopefully in 1.20?)